File: 0006 Zigzag Conversion/ZigzagConversion.go
package main

import (
	"bytes"
)

func main() {
	convert("PAYPALISHIRING", 1)
}

func convert(s string, numRows int) string {
	len := len(s)
	if len <= numRows || numRows < 2 {
		return s
	}

	var ret bytes.Buffer
	for r := 0; r < numRows; r++ {
		index := r
		offset0 := 2 * (numRows - r - 1)
		offset1 := 2 * r

		ret.WriteString(string(s[index]))
		for index < len {
			if offset0 > 0 {
				index += offset0
				if index < len {
					ret.WriteString(string(s[index]))
				}
			}
			if offset1 > 0 {
				index += offset1
				if index < len {
					ret.WriteString(string(s[index]))
				}
			}
		}
	}
	return ret.String()
}

File: 0006 Zigzag Conversion/ZigzagConversion.cs
public class Solution {
    public string Convert(string s, int numRows) {
        int len = s.Length;
        if (len <= numRows || numRows < 2) return s;

        string[] ret = new string[len];
        int i = 0;
        for (int r = 0; r < numRows; r++) {
            int index = r;
            int offset0 = 2 * (numRows - r - 1);
            int offset1 = 2 * r;

            ret[i++] += s[index];
            for (; index < len;) {
                if (offset0 > 0) {
                    index += offset0;
                    if (index < len) {
                        ret[i++] += s[index];
                    }
                }
                if (offset1 > 0) {
                    index += offset1;
                    if (index < len) {
                        ret[i++] += s[index];
                    }
                }
            }
        }
        return string.Join("", ret);
    }
}